gadget/install/encrypt.go

```diff
@@ -88,3 +105,66 @@ func cryptsetupClose(name string) error {
 	}
 	return nil
 }
+
+// encryptedDeviceWithSetupHook represents a block device that is setup using
+// the "device-setup" hook.
+type encryptedDeviceWithSetupHook struct {
+	parent *gadget.OnDiskStructure
+	name   string
+	node   string
+}
+
+// sanity
+var _ = encryptedDevice(&encryptedDeviceWithSetupHook{})
+
+// newEncryptedDeviceWithSetupHook creates an encrypted device in the
+// existing partition using the specified key using the fde-setup hook
+func newEncryptedDeviceWithSetupHook(part *gadget.OnDiskStructure, key secboot.EncryptionKey, name string) (encryptedDevice, error) {
+	// 1. create linear mapper device with 1Mb of reserved space
+	uuid := ""
+	offset := fde.DeviceSetupHookPartitionOffset
+	sizeMinusOffset := uint64(part.Size) - offset
+	mapperDevice, err := disks.CreateLinearMapperDevice(part.Node, name, uuid, offset, sizeMinusOffset)
+	if err != nil {
+		return nil, err
+	}
+
+	// 2. run fde-setup "device-setup" on it
+	// TODO: We may need a different way to run the fde-setup hook
+	//       here. The hook right now runs with a locked state. But
+	//       when this runs the state will be unlocked but our hook
+	//       mechanism needs a locked state. This means we either need
+	//       something like "boot.RunFDE*Device*SetupHook" or we run
+	//       the entire install with the state locked (which may not
+	//       be as terrible as it sounds as this is a rare situation).
+	runHook := boot.RunFDEDeviceSetupHook
+	params := &fde.DeviceSetupParams{
+		Key:    key,
+		Device: mapperDevice,
+		Label:  part.Label,
```

Comment:
This needs to be more specific: either "PartitionLabel" or "FilesystemLabel" or the same as fdisk

Reply:
Yeah, a more descriptive name would be useful here. Also, it isn't which device this label is associated with?

Reply:
I've renamed this to ParitioinName as that's how it's called for GPT. AFAIU we do not need to pass filesystem label, as that is known only later after unlocking the device?

Reply:
Indeed, that should be enough - once renamed @kubiko needs to update his hook code to look at the new more precise name